File: src/cli/show.test.ts
import { writeFileSync } from "node:fs";
import { join } from "node:path";
import { test } from "node:test";

import { expect } from "expect";

import TempDir from "../TempDir.ts";
import { type Advisory, type AuditSource } from "../audit/index.ts";
import { loadConfig } from "../config.ts";
import {
  type Coordinates,
  type CoordinateString,
  coordinatesToString,
  InMemoryPackageSource,
  type PackageMetadata,
  toCoordinates,
} from "../packages/index.ts";
import { buildShowData, renderShowCard, type ShowData, showToJson } from "./show.ts";

// A no-op paint so the card text asserts on content, not ANSI codes.
const plain = (_format: unknown, text: string): string => text;

function source(): InMemoryPackageSource {
  const pkg = (
    g: string,
    a: string,
    v: string,
    extra: Partial<PackageMetadata> = {},
  ): PackageMetadata => ({ coordinates: toCoordinates(g, a, v), dependencies: [], ...extra });
  return new InMemoryPackageSource("test", [
    pkg("com.google.code.gson", "gson", "2.11.0"),
    pkg("com.google.code.gson", "gson", "2.13.0"),
    pkg("com.google.code.gson", "gson", "2.13.1", {
      description: "A library to convert Java Objects into JSON and back",
      homepage: "https://github.com/google/gson",
      scmUrl: "https://github.com/google/gson.git",
      licenses: [{ name: "Apache-2.0", url: "https://www.apache.org/licenses/LICENSE-2.0.txt" }],
      dependencies: [toCoordinates("com.google.errorprone", "error_prone_annotations", "2.27.0")],
    }),
  ]);
}

/** An AuditSource that returns canned advisories for specific coordinates. */
function auditStub(byCoord: Record<string, Advisory[]> = {}): AuditSource {
  return {
    name: "stub",
    findVulnerabilities(coordinates: readonly Coordinates[]) {
      const out = new Map<CoordinateString, Advisory[]>();
      for (const c of coordinates) {
        const key = coordinatesToString(c);
        if (byCoord[key]) out.set(key, byCoord[key]);
      }
      return Promise.resolve(out);
    },
  };
}

function configWith(json: object): ReturnType<typeof loadConfig> {
  const dir = TempDir.create("cappu-show-");
  writeFileSync(join(dir.path, "cappu.json"), JSON.stringify(json));
  return loadConfig(undefined, dir.path);
}

async function ok(
  coord: string,
  config: ReturnType<typeof loadConfig>,
  audit: AuditSource = auditStub(),
): Promise<ShowData> {
  const data = await buildShowData(coord, config, [source()], audit);
  if ("error" in data) throw new Error(`unexpected error: ${data.error}`);
  return data;
}

test("show defaults to the latest version and reports it as latest", async () => {
  const config = configWith({});
  const data = await ok("com.google.code.gson:gson", config);

  expect(data.version).toBe("2.13.1");
  expect(data.latestVersion).toBe("2.13.1");
  expect(data.versionCount).toBe(3);
  expect(data.explicitVersion).toBe(false);
  expect(data.spdx).toEqual(["Apache-2.0"]);
  expect(data.homepage).toBe("https://github.com/google/gson");
  expect(data.project.configurations).toEqual([]);

  const card = renderShowCard(data, plain);
  expect(card).toContain("com.google.code.gson:gson 2.13.1  latest");
  expect(card).toContain("convert Java Objects into JSON");
  expect(card).toContain("License      Apache-2.0");
  expect(card).toContain("Homepage     https://github.com/google/gson");
  expect(card).toContain("Repository   https://github.com/google/gson.git");
  expect(card).toContain("not a direct dependency");
  expect(card).toContain("Dependencies (1)");
  expect(card).toContain("com.google.errorprone:error_prone_annotations:2.27.0");
  expect(card).toContain("no known vulnerabilities");
});

test("show on an older pinned version flags newer releases", async () => {
  const config = configWith({});
  const data = await ok("com.google.code.gson:gson:2.11.0", config);

  expect(data.version).toBe("2.11.0");
  expect(data.explicitVersion).toBe(true);
  expect(data.newer).toBe(2);
  expect(renderShowCard(data, plain)).toContain("2.11.0  2 newer available");
});

test("show surfaces how this project depends on the package", async () => {
  const config = configWith({
    dependencies: { implementation: { "com.google.code.gson:gson": "2.13.1" } },
  });
  const data = await ok("com.google.code.gson:gson", config);

  expect(data.project.configurations).toEqual(["implementation"]);
  expect(data.project.declared).toBe("2.13.1");
  expect(renderShowCard(data, plain)).toContain("In project   implementation (declared 2.13.1)");
});

test("show reports known vulnerabilities and exits non-zero", async () => {
  const config = configWith({});
  const advisory: Advisory = {
    id: "GHSA-xxxx-yyyy-zzzz" as Advisory["id"],
    aliases: ["CVE-2022-0001"],
    summary: "Example deserialization issue",
    severity: "high",
    fixedVersions: ["2.13.1"],
    url: "https://osv.dev/vulnerability/GHSA-xxxx-yyyy-zzzz",
  };
  const audit = auditStub({ "com.google.code.gson:gson:2.13.0": [advisory] });
  const data = await ok("com.google.code.gson:gson:2.13.0", config, audit);

  expect(data.vulnerabilities).toHaveLength(1);
  const card = renderShowCard(data, plain);
  expect(card).toContain(
    "HIGH  GHSA-xxxx-yyyy-zzzz (CVE-2022-0001) - Example deserialization issue",
  );
  expect(card).toContain("[fixed in: 2.13.1]");
});

test("show --json carries the same data", async () => {
  const config = configWith({
    dependencies: { implementation: { "com.google.code.gson:gson": "2.13.1" } },
  });
  const data = await ok("com.google.code.gson:gson", config);
  const json = showToJson(data) as Record<string, unknown>;

  expect(json.version).toBe("2.13.1");
  expect(json.latestVersion).toBe("2.13.1");
  expect(json.license).toEqual(["Apache-2.0"]);
  expect(json.homepage).toBe("https://github.com/google/gson");
  expect(json.dependencies).toEqual([
    {
      groupId: "com.google.errorprone",
      artifactId: "error_prone_annotations",
      version: "2.27.0",
    },
  ]);
  expect(json.project).toEqual({
    configurations: ["implementation"],
    declared: "2.13.1",
    installed: null,
  });
  expect(json.vulnerabilities).toEqual([]);
});

test("show errors on a malformed coordinate and an unknown package", async () => {
  const config = configWith({});
  expect(await buildShowData("not-a-coord", config, [source()], auditStub())).toEqual({
    error:
      "show needs group:artifact[:version], e.g. `cappu show com.google.code.gson:gson`; " +
      "search for a package with `cappu search <query>`",
    code: 2,
  });
  expect(await buildShowData("org.x:nope", config, [source()], auditStub())).toEqual({
    error: "package not found: org.x:nope; search for a package with `cappu search <query>`",
    code: 1,
  });
});

File: togo/internal/cli/show_test.go
package cli

import (
	"os"
	"path/filepath"
	"strings"
	"testing"

	"github.com/nikeee/cappu/internal/audit"
	"github.com/nikeee/cappu/internal/config"
	"github.com/nikeee/cappu/internal/packages"
)

// plain is a no-op paint so the card asserts on content, not ANSI codes.
func plain(_ string, text string) string { return text }

// showAuditStub returns canned advisories keyed by "group:artifact:version".
type showAuditStub struct {
	byCoord map[packages.CoordinateString][]audit.Advisory
}

func (showAuditStub) Name() string { return "stub" }
func (s showAuditStub) FindVulnerabilities(coords []packages.Coordinates) (map[packages.CoordinateString][]audit.Advisory, error) {
	out := map[packages.CoordinateString][]audit.Advisory{}
	for _, c := range coords {
		if adv, ok := s.byCoord[c.String()]; ok {
			out[c.String()] = adv
		}
	}
	return out, nil
}

func showSource() *packages.InMemoryPackageSource {
	pkg := func(g, a, v string, mut func(*packages.PackageMetadata)) packages.PackageMetadata {
		m := packages.PackageMetadata{Coordinates: packages.NewCoordinates(g, a, v)}
		if mut != nil {
			mut(&m)
		}
		return m
	}
	return packages.NewInMemoryPackageSource("test", []packages.PackageMetadata{
		pkg("com.google.code.gson", "gson", "2.11.0", nil),
		pkg("com.google.code.gson", "gson", "2.13.0", nil),
		pkg("com.google.code.gson", "gson", "2.13.1", func(m *packages.PackageMetadata) {
			m.Description = "A library to convert Java Objects into JSON and back"
			m.Homepage = "https://github.com/google/gson"
			m.ScmURL = "https://github.com/google/gson.git"
			m.Licenses = []packages.License{{Name: "Apache-2.0", URL: "https://www.apache.org/licenses/LICENSE-2.0.txt"}}
			m.Dependencies = []packages.DependencyDeclaration{
				{Coordinates: packages.NewCoordinates("com.google.errorprone", "error_prone_annotations", "2.27.0")},
			}
		}),
	})
}

func showConfig(t *testing.T, cappuJSON string) *config.Config {
	t.Helper()
	dir := t.TempDir()
	if cappuJSON == "" {
		cappuJSON = "{}"
	}
	if err := os.WriteFile(filepath.Join(dir, "cappu.json"), []byte(cappuJSON), 0o644); err != nil {
		t.Fatal(err)
	}
	cfg, err := config.Load("", dir)
	if err != nil {
		t.Fatal(err)
	}
	return cfg
}

func mustShow(t *testing.T, coord string, cfg *config.Config, a audit.AuditSource) *showData {
	t.Helper()
	if a == nil {
		a = showAuditStub{}
	}
	data, showErr := buildShowData(coord, cfg, []packages.PackageSource{showSource()}, a)
	if showErr != nil {
		t.Fatalf("unexpected error: %s", showErr.message)
	}
	return data
}

func TestShowDefaultsToLatest(t *testing.T) {
	data := mustShow(t, "com.google.code.gson:gson", showConfig(t, ""), nil)
	if data.version != "2.13.1" || data.latestVersion != "2.13.1" || data.versionCount != 3 {
		t.Fatalf("version=%q latest=%q count=%d", data.version, data.latestVersion, data.versionCount)
	}
	if data.explicitVersion {
		t.Error("explicitVersion = true, want false")
	}
	card := renderShowCard(data, plain)
	for _, want := range []string{
		"com.google.code.gson:gson 2.13.1  latest",
		"convert Java Objects into JSON",
		"License      Apache-2.0",
		"Homepage     https://github.com/google/gson",
		"Repository   https://github.com/google/gson.git",
		"not a direct dependency",
		"Dependencies (1)",
		"com.google.errorprone:error_prone_annotations:2.27.0",
		"no known vulnerabilities",
	} {
		if !strings.Contains(card, want) {
			t.Errorf("card missing %q\n%s", want, card)
		}
	}
}

func TestShowOlderPinnedFlagsNewer(t *testing.T) {
	data := mustShow(t, "com.google.code.gson:gson:2.11.0", showConfig(t, ""), nil)
	if !data.explicitVersion || data.newer != 2 {
		t.Fatalf("explicit=%v newer=%d", data.explicitVersion, data.newer)
	}
	if card := renderShowCard(data, plain); !strings.Contains(card, "2.11.0  2 newer available") {
		t.Errorf("card missing newer hint\n%s", card)
	}
}

func TestShowProjectContext(t *testing.T) {
	cfg := showConfig(t, `{"dependencies":{"implementation":{"com.google.code.gson:gson":"2.13.1"}}}`)
	data := mustShow(t, "com.google.code.gson:gson", cfg, nil)
	if len(data.project.configurations) != 1 || data.project.configurations[0] != "implementation" {
		t.Fatalf("configurations = %v", data.project.configurations)
	}
	if data.project.declared != "2.13.1" {
		t.Errorf("declared = %q", data.project.declared)
	}
	if card := renderShowCard(data, plain); !strings.Contains(card, "In project   implementation (declared 2.13.1)") {
		t.Errorf("card missing project row\n%s", card)
	}
}

func TestShowVulnerabilities(t *testing.T) {
	stub := showAuditStub{byCoord: map[packages.CoordinateString][]audit.Advisory{
		"com.google.code.gson:gson:2.13.0": {{
			ID:            "GHSA-xxxx-yyyy-zzzz",
			Aliases:       []string{"CVE-2022-0001"},
			Summary:       "Example deserialization issue",
			Severity:      audit.SeverityHigh,
			FixedVersions: []string{"2.13.1"},
			URL:           "https://osv.dev/vulnerability/GHSA-xxxx-yyyy-zzzz",
		}},
	}}
	data := mustShow(t, "com.google.code.gson:gson:2.13.0", showConfig(t, ""), stub)
	if len(data.vulnerabilities) != 1 {
		t.Fatalf("vulnerabilities = %d", len(data.vulnerabilities))
	}
	card := renderShowCard(data, plain)
	if !strings.Contains(card, "HIGH  GHSA-xxxx-yyyy-zzzz (CVE-2022-0001) - Example deserialization issue") {
		t.Errorf("card missing advisory line\n%s", card)
	}
	if !strings.Contains(card, "[fixed in: 2.13.1]") {
		t.Errorf("card missing fixed hint\n%s", card)
	}
}

func TestShowJSON(t *testing.T) {
	cfg := showConfig(t, `{"dependencies":{"implementation":{"com.google.code.gson:gson":"2.13.1"}}}`)
	data := mustShow(t, "com.google.code.gson:gson", cfg, nil)
	j := showToJSON(data)
	if j.Version != "2.13.1" || j.LatestVersion == nil || *j.LatestVersion != "2.13.1" {
		t.Fatalf("version=%q latest=%v", j.Version, j.LatestVersion)
	}
	if len(j.License) != 1 || j.License[0] != "Apache-2.0" {
		t.Errorf("license = %v", j.License)
	}
	if len(j.Dependencies) != 1 || j.Dependencies[0].ArtifactID != "error_prone_annotations" {
		t.Errorf("dependencies = %+v", j.Dependencies)
	}
	if len(j.Project.Configurations) != 1 || j.Project.Declared == nil || *j.Project.Declared != "2.13.1" {
		t.Errorf("project = %+v", j.Project)
	}
	if j.Project.Installed != nil {
		t.Errorf("installed = %v, want nil", j.Project.Installed)
	}
	if len(j.Vulnerabilities) != 0 {
		t.Errorf("vulnerabilities = %v", j.Vulnerabilities)
	}
}

func TestShowErrors(t *testing.T) {
	cfg := showConfig(t, "")
	if _, e := buildShowData("not-a-coord", cfg, []packages.PackageSource{showSource()}, showAuditStub{}); e == nil || e.code != 2 {
		t.Fatalf("malformed coord error = %v", e)
	}
	if _, e := buildShowData("org.x:nope", cfg, []packages.PackageSource{showSource()}, showAuditStub{}); e == nil || e.code != 1 {
		t.Fatalf("unknown package error = %v", e)
	}
}
